### PR TITLE
make a more robust sample

### PR DESCRIPTION
**Pull request by sweetcoco**

### CODE REVIEW

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 1]
```diff
@@ -54,25 +55,65 @@
...
+                String payload = IOUtils.toString(response.getEntity().getContent(), "UTF-8");
+                logger.debug("Response: " + payload);
+                if (response.getStatusLine().getStatusCode() >= 300) {
+                    throw new HttpResponseException(response.getStatusLine().getStatusCode(),
+                            response.getStatusLine().getReasonPhrase());
+                }
+                return payload;
+            } catch (Exception e) {
+                logger.error("Something unexpected happened: " + e.getMessage());
+                TokenStore.delete(username); // fixme - needed only for 401
```
Comment (ohmslaw) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco I see a few `fixme` comments?

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 2]
```diff
@@ -54,25 +55,65 @@
...
+            }
+        }
+    }
+
+    @RequestMapping(value = "/loop-template", method = RequestMethod.GET)
+    @ResponseBody
+    public String getLoopTemplates(@RequestParam(value = "profile_id", required = true) String profile_id) {
 
-        Token token = TokenStore.get(user.getUsername());
+        Token token = TokenStore.get(username);
```
Comment (ohmslaw) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco Should we have a method to handle this duplicate code including the conditional below?

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 3]
```diff
@@ -82,23 +123,29 @@ public String loopIt() {
...
     }
 
-    @RequestMapping("/hello")
-    public String hello(HttpServletRequest request, Model model) {
-        User user = getUser();
+    @RequestMapping("/")
+    public String home(HttpServletRequest request, Model model) {
+//        User user = getUser();
+
+        boolean connected = !StringUtils.isEmpty(TokenStore.get(username));
```
Comment (ohmslaw) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco Is there `isNotEmpty`?

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 4]
```diff
@@ -82,23 +123,29 @@ public String loopIt() {
...
-        model.addAttribute("username", StringUtils.isEmpty(TokenStore.get(user.getUsername())));
+        model.addAttribute("authorize_url", getAuthorizeUrl());
+        model.addAttribute("username", StringUtils.isEmpty(TokenStore.get(username)));
+
+        if (connected) {
+            model.addAttribute("profiles", getProfiles());
+        }
 
-        return "hello";
+        return "home";
```
Comment (ohmslaw) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco home?

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 5]
```diff
@@ -109,39 +156,69 @@ public void deleteToken() {
...
+                if (response.getStatusLine().getStatusCode() >= 300) {
+                    throw new HttpResponseException(response.getStatusLine().getStatusCode(),
+                            response.getStatusLine().getReasonPhrase());
+                }
+
+                return payload;
+            } catch (Exception e) {
+                logger.error("Something unexpected happened: " + e.getMessage());
+                TokenStore.delete(username); // fixme - needed only for 401
+                throw new RuntimeException(e); // todo handle error - token revoked, etc
```
Comment (ohmslaw) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco This almost all feels like a duplicate of the code above?

File: src/main/resources/templates/home.html [hunk 6]
```diff
@@ -1,11 +1,574 @@
...
+            profiles: [],
+            step: 'connect',
+            participants: [],
+            $participants: [],
+            $loopForm: null
+        };
+
+        var steps = ['connect', 'select-profile', 'select-loop-template', 'loop-it'];
+
+        var token = $("meta[name='_csrf']").attr("content");
```
Comment (ohmslaw) [inline src/main/resources/templates/home.html:31]: @sweetcoco I thought I saw the csrf code in the javas commented out?

File: src/main/resources/templates/home.html [hunk 7]
```diff
@@ -1,11 +1,574 @@
...
+        var header = $("meta[name='_csrf_header']").attr("content");
+        var headers = {};
+        headers[header] = token;
+
+        function popupwindow(url, title, w, h) {
+            var left = (screen.width/2)-(w/2);
+            var top = (screen.height/2)-(h/2);
```
Comment (ohmslaw) [inline src/main/resources/templates/home.html:38]: @sweetcoco Did you verify this is functional in a few browsers? Does not always need to be pretty, but it should be functional.

File: src/main/resources/templates/home.html [hunk 8]
```diff
@@ -1,11 +1,574 @@
...
+                    state.profiles = JSON.parse(state.JSONprofiles).data;
+                }
+
+                if (!state.selectedLoopTemplate) {
+                    state.step = 'select-loop-template';
+                }
+
+                if (!state.selectedProfile) {
+                    state.step = 'select-profile';
+                }
```
Comment (ohmslaw) [inline src/main/resources/templates/home.html:71]: @sweetcoco Should these be more of an `if-else` kind of deal?

File: src/main/resources/templates/home.html [hunk 9]
```diff
@@ -1,11 +1,574 @@
...
+
+            renderParticipants();
+        };
+
+        var initListeners = function() {
+
+            $("#connect").click(function () {
+                var win = popupwindow(/*[[${authorize_url}]]*/, "Connect with Dotloop!", 800, 600);
+                checkConnect = setInterval(function() {
+                    if (!win || !win.closed) return;
```
Comment (ohmslaw) [inline src/main/resources/templates/home.html:172]: @sweetcoco ...

**Review by ohmslaw — COMMENTED** [review]
Few comments but overall it is not too bad to consume. This does feel very verbose and might be hard to maintain if you were planning many iterations but as a one-off we do not expect to change much, I think it is decent.

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 10]
```diff
@@ -54,51 +51,102 @@
     @Value("${dotloop.api.endpoint}")
     private String apiEndpoint;
 
-    @RequestMapping("/loopit")
+    private String username = "api user";
```
Comment (adunn8) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco  convention for string constants is ALL_CAPS_WITH_UNDERSCORES

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 11]
```diff
@@ -54,51 +51,102 @@
...
+        return request;
+    }
+
+    private String makeRequest(String type, String params, String bodyData) throws Exception {
+        try {
+            params = params.isEmpty() ? "" : params;
+
+            Request request = attachHeaders(Request.Get(getApiBaseUrl() + params));
+
+            if (type.equalsIgnoreCase("post")) {
```
Comment (adunn8) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco instead of creating conditional logic on the http method,  a better approach might be to create dedicated method for each request type (ie getRequest, postRequest, ect.).  Since the execution and response processing is the same, you could just have each method use the same shared logic.  That would ensure that the calling code is declaring their request type without having to worry that they are setting the "type" parameter correctly.  Also mirrors the same pattern as the underlying library.

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 12]
```diff
@@ -54,51 +51,102 @@
...
+
+    private Request attachBody(Request request,  String bodyData) {
+        request.bodyString(bodyData, ContentType.APPLICATION_JSON);
+
+        return request;
+    }
+
+    private String makeRequest(String type, String params, String bodyData) throws Exception {
+        try {
+            params = params.isEmpty() ? "" : params;
```
Comment (adunn8) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco rather than resetting the param object, might be cleaner if you just returned the actual URL

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 13]
```diff
@@ -54,51 +51,102 @@
...
+
+            Request request = attachHeaders(Request.Get(getApiBaseUrl() + params));
+
+            if (type.equalsIgnoreCase("post")) {
+                request = attachBody(attachHeaders(Request.Post(getApiBaseUrl() + params)), bodyData);
```
Comment (adunn8) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco this is a little hard to parse, ideally we'd have an object and then call a series of actions on it , where as this is the inverse of that pattern.  If you use the suggestion of creating dedicated request type methods, you can just attach the body directly to the Request object (`attachBody` is basically 1:1 swap code) and then set your headers in the shared request processing code.

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 14]
```diff
@@ -54,51 +51,102 @@
     @Value("${dotloop.api.endpoint}")
     private String apiEndpoint;
 
-    @RequestMapping("/loopit")
+    private String username = "api user";
+
+    private Request attachHeaders(Request request) {
+        Token token = TokenStore.get(username);
+        request.addHeader("Authorization", "Bearer " + token.getAccessToken())
+                .addHeader("Content-type", "application/json");
```
Comment (adunn8) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco prob makes sense to use `ContentType.APPLICATION_JSON` constant here as well

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 15]
```diff
@@ -54,51 +51,102 @@
...
-    public String loopIt() {
-        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+    public String loopIt(
+            @RequestParam(value = "profile_id", required = true) String profile_id,
+            @RequestBody Loop loop) throws Exception {
+
+        Token token = TokenStore.get(username);
 
-        Token token = TokenStore.get(user.getUsername());
+        ObjectMapper mapper = new ObjectMapper();
```
Comment (adunn8) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco can you make this a member variable?  In many cases the object mapper will have customized serialization settings that need to be shared.  Not really the case here, but good practice.

File: src/main/java/com/dotloop/loopit/LoopItController.java [hunk 16]
```diff
@@ -109,39 +157,43 @@ public void deleteToken() {
...
+
     @ResponseStatus(HttpStatus.FORBIDDEN)
     public class ForbiddenException extends RuntimeException {
         public ForbiddenException(String msg) {
             super(msg);
         }
     }
 
-    private String getAuthorizeUrl(String state) {
+    private String getAuthorizeUrl() {
```
Comment (adunn8) [inline src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco any reason that we removed state?  We recommend that our users implement this in our developer guide, so it might be a good idea to include in the sample app if not too difficult

**Review by adunn8 — CHANGES_REQUESTED** [review]
@sweetcoco added a few comments, mostly around cleaning up the logic for making public api requests

File: src/main/java/com/dotloop/loopit/LoopItController.java (hunk 13 above)
Comment (adunn8) [reply src/main/java/com/dotloop/loopit/LoopItController.java]: @sweetcoco Here's some (untested) code to give you an idea of what I was looking for:

```
    private String post(String url, String bodyData){

        Request postRequest = Request.Post(getApiBaseUrl() + url).bodyString(bodyData, ContentType.APPLICATION_JSON);
        return executeRequest(postRequest);
        
    }

    private String executeRequest(Request request){

        try {

            Token token = TokenStore.get(username);
            request.addHeader("Authorization", "Bearer " + token.getAccessToken())
                    .addHeader("Content-type", "application/json");


            HttpResponse response = request.execute().returnResponse();

            String payload = IOUtils.toString(response.getEntity().getContent(), "UTF-8");
            logger.debug("Response: " + payload);
            if (response.getStatusLine().getStatusCode() >= 300) {
                throw new HttpResponseException(response.getStatusLine().getStatusCode(),
                        response.getStatusLine().getReasonPhrase());
            }

            return payload;

        } catch (Exception e) {
            logger.error("Something unexpected happened: " + e.getMessage());
            TokenStore.delete(username); // fixme - needed only for 401
            throw new RuntimeException(e); // todo handle error - token revoked, etc
        }

    }
```

Comment (sweetcoco) [reply src/main/java/com/dotloop/loopit/LoopItController.java]: @adunn8 you really went above and beyond with this comment, thanks man.

**Review by adunn8 — APPROVED** [review]
@sweetcoco approved!
